In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import json # library to handle JSON files

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [121]:
#loading html file as a data frame
table = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M',header=0)


In [3]:
df=table[0] # we only need the first table
print(type(df))# checking if df is a data frame
print(df.shape)
df.head()

<class 'pandas.core.frame.DataFrame'>
(180, 3)


Postal code           Borough                Neighborhood
0         M1A      Not assigned                         NaN
1         M2A      Not assigned                         NaN
2         M3A        North York                   Parkwoods
3         M4A        North York            Victoria Village
4         M5A  Downtown Toronto  Regent Park / Harbourfront

# Data cleaning
### We need to only consider the rows with an assigned Borough


In [4]:
df['Borough'].replace('Not assigned',np.NaN,inplace=True)
df.isnull().sum()

Postal code      0
Borough         77
Neighborhood    77
dtype: int64

So there are 77 rows without Boroughs, We can drop these rows

In [5]:
df.dropna(axis=0,subset=['Borough'],inplace=True)

df.head()


Postal code           Borough                                  Neighborhood
2         M3A        North York                                     Parkwoods
3         M4A        North York                              Victoria Village
4         M5A  Downtown Toronto                    Regent Park / Harbourfront
5         M6A        North York             Lawrence Manor / Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

We can verify whether the drop is succesfull

In [6]:
df.isnull().sum()

Postal code     0
Borough         0
Neighborhood    0
dtype: int64

Yes the rows are correct. let us also reset the indexes

In [7]:
df=df.reset_index(drop=True)
df.head()

Postal code           Borough                                  Neighborhood
0         M3A        North York                                     Parkwoods
1         M4A        North York                              Victoria Village
2         M5A  Downtown Toronto                    Regent Park / Harbourfront
3         M6A        North York             Lawrence Manor / Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

In [8]:
df.shape

(103, 3)

Data Frame is having 103 rows and 3 columns

# Geocoding the data
### Since Gecode is not persitent I choose to load the csv file

In [9]:
pos = pd.read_csv('https://cocl.us/Geospatial_data')
pos.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [10]:
pos.rename(columns={'Postal Code':'Postal code'},inplace=True) # making both dataframes have same column label

Now we cam use pd.merge() to join two dataframes into a new dataframe Toronto_Neighborhood

In [11]:
Toronto_Neighborhood = pd.merge(df,pos,on='Postal code')
Toronto_Neighborhood.head()

Postal code           Borough                                  Neighborhood  \
0         M3A        North York                                     Parkwoods   
1         M4A        North York                              Victoria Village   
2         M5A  Downtown Toronto                    Regent Park / Harbourfront   
3         M6A        North York             Lawrence Manor / Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

# Exploring neighborhood using Foursquare

In [13]:
def getNearbyVenues(names, latitudes, longitudes, radius=500,LIMIT=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)   
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [14]:
toronto_venues = getNearbyVenues(names=Toronto_Neighborhood['Neighborhood'],
                                   latitudes=Toronto_Neighborhood['Latitude'],
                                   longitudes=Toronto_Neighborhood['Longitude'])


Parkwoods
Victoria Village
Regent Park / Harbourfront
Lawrence Manor / Lawrence Heights
Queen's Park / Ontario Provincial Government
Islington Avenue
Malvern / Rouge
Don Mills
Parkview Hill / Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park / Princess Gardens / Martin Grove / Islington / Cloverdale
Rouge Hill / Port Union / Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate / Bloordale Gardens / Old Burnhamthorpe / Markland Wood
Guildwood / Morningside / West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor / Wilson Heights / Downsview North
Thorncliffe Park
Richmond / Adelaide / King
Dufferin / Dovercourt Village
Scarborough Village
Fairview / Henry Farm / Oriole
Northwood Park / York University
East Toronto
Harbourfront East / Union Station / Toronto Islands
Little Portugal / Trinity
Kennedy Park / Ionview / East Birchmount Park
Bayview Village
Do

Resultant data frame

In [15]:
print(toronto_venues.shape)
toronto_venues.head()

(2129, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                               Venue  Venue Latitude  Venue Longitude  \
0                    Brookbanks Park       43.751976       -79.332140   
1                      Variety Store       43.751974       -79.333114   
2  Corrosion Service Company Limited       43.752432       -79.334661   
3             Victoria Village Arena       43.723481       -79.315635   
4                        Tim Hortons       43.725517       -79.313103   

               Venue Category  
0                        Park  
1           Food & Drink Shop  
2  Construction & Landscaping  
3                Hockey Arena  
4                 Coffee Shop

Venue count of each neighborhood

In [16]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               3   
Alderwood / Long Branch                                                 8   
Bathurst Manor / Wilson Heights / Downsview North                      21   
Bayview Village                                                         4   
Bedford Park / Lawrence Manor East                                     24   
Berczy Park                                                            58   
Birch Cliff / Cliffside West                                            4   
Brockton / Parkdale Village / Exhibition Place                         23   
Business reply mail Processing CentrE                                  16   
CN Tower / King and Spadina / Railway Lands / H...                     18   
Caledonia-Fairbanks                                                     4   
Canada Post Gateway Processing Centre                                  11   
Cedarbrae                                                               8   
Central Bay Street                                                     64   
Christie                                                               18   
Church and Wellesley                                                   76   
Clarks Corners / Tam O'Shanter / Sullivan                              15   
Cliffside / Cliffcrest / Scarborough Village West                       2   
Commerce Court / Victoria Hotel                                       100   
Davisville                                                             37   
Davisville North                                                        8   
Del Ray / Mount Dennis / Keelsdale and Silverthorn                      4   
Don Mills                                                              26   
Dorset Park / Wexford Heights / Scarborough Tow...                      5   
Downsview                                                              14   
Dufferin / Dovercourt Village                                          20   
East Toronto                                                            2   
Eringate / Bloordale Gardens / Old Burnhamthorp...                      8   
Fairview / Henry Farm / Oriole                                         62   
First Canadian Place / Underground city                               100   
Forest Hill North & West                                                5   
Garden District, Ryerson                                              100   
Glencairn                                                               4   
Golden Mile / Clairlea / Oakridge                                       9   
Guildwood / Morningside / West Hill                                     7   
Harbourfront East / Union Station / Toronto Isl...                    100   
High Park / The Junction South                                         24   
Hillcrest Village                                                       4   
Humber Summit                                                           1   
Humberlea / Emery                                                       2   
Humewood-Cedarvale                                                      4   
India Bazaar / The Beaches West                                        19   
Kennedy Park / Ionview / East Birchmount Park                           3   
Kensington Market / Chinatown / Grange Park                            55   
Kingsview Village / St. Phillips / Martin Grove...                      4   
Lawrence Manor / Lawrence Heights                                      14   
Lawrence Park                                                           3   
Leaside                                                                33   
Little Portugal / Trinity                                              41   
Malvern / Rouge                                                         1   
Milliken / Agincourt North / Steeles East / L'A.

Find out how many unique venues are there


In [17]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 269 uniques categories.


# Analyse neighbour hood and their venues

In [18]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                           Agincourt     0.000000   
1                             Alderwood / Long Branch     0.000000   
2   Bathurst Manor / Wilson Heights / Downsview North     0.000000   
3                                     Bayview Village     0.000000   
4                  Bedford Park / Lawrence Manor East     0.000000   
5                                         Berczy Park     0.000000   
6                        Birch Cliff / Cliffside West     0.000000   
7      Brockton / Parkdale Village / Exhibition Place     0.000000   
8               Business reply mail Processing CentrE     0.000000   
9   CN Tower / King and Spadina / Railway Lands / ...     0.000000   
10                                Caledonia-Fairbanks     0.000000   
11              Canada Post Gateway Processing Centre     0.000000   
12                                          Cedarbrae     0.000000   
13                                 Central Bay Street     0.015625   
14                                           Christie     0.000000   
15                               Church and Wellesley     0.026316   
16          Clarks Corners / Tam O'Shanter / Sullivan     0.000000   
17  Cliffside / Cliffcrest / Scarborough Village West     0.000000   
18                    Commerce Court / Victoria Hotel     0.000000   
19                                         Davisville     0.000000   
20                                   Davisville North     0.000000   
21  Del Ray / Mount Dennis / Keelsdale and Silvert...     0.000000   
22                                          Don Mills     0.000000   
23  Dorset Park / Wexford Heights / Scarborough To...     0.000000   
24                                          Downsview     0.000000   
25                      Dufferin / Dovercourt Village     0.000000   
26                                       East Toronto     0.000000   
27  Eringate / Bloordale Gardens / Old Burnhamthor...     0.000000   
28                     Fairview / Henry Farm / Oriole     0.000000   
29            First Canadian Place / Underground city     0.000000   
30                           Forest Hill North & West     0.000000   
31                           Garden District, Ryerson     0.000000   
32                                          Glencairn     0.000000   
33                  Golden Mile / Clairlea / Oakridge     0.000000   
34                Guildwood / Morningside / West Hill     0.000000   
35  Harbourfront East / Union Station / Toronto Is...     0.000000   
36                     High Park / The Junction South     0.000000   
37                                  Hillcrest Village     0.000000   
38                                      Humber Summit     0.000000   
39                                  Humberlea / Emery     0.000000   
40                                 Humewood-Cedarvale     0.000000   
41                    India Bazaar / The Beaches West     0.000000   
42      Kennedy Park / Ionview / East Birchmount Park     0.000000   
43        Kensington Market / Chinatown / Grange Park     0.000000   
44  Kingsview Village / St. Phillips / Martin Grov...     0.000000   
45                  Lawrence Manor / Lawrence Heights     0.000000   
46                                      Lawrence Park     0.000000   
47                                            Leaside     0.000000   
48                          Little Portugal / Trinity     0.024390   
49                                    Malvern / Rouge     0.000000   
50  Milliken / Agincourt North / Steeles East / L'...     0.000000   
51  Mimico NW / The Queensway West / South of Bloo...     0.000000   
52                       Moore Park / Summerhill East     0.000000   
53     New Toronto / Mimico South / Humber Bay Shores     0.000000   
54         North Park / Maple Leaf Park / Upwood Park     0.000000   
55                                 North Toronto West     0.047619   
56                                          Northwe

we have a detailed dataframe about each neighborhood shape of the dataframe is:

In [19]:
print(toronto_grouped.shape)

(93, 269)


Now we can create a dataframe to display 10 most common venues, so we can cluster them on the basis of similarity

In [89]:
#function for sorting venues in a descending order
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]
# top 10 venues
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  \
0                                          Agincourt   
1                            Alderwood / Long Branch   
2  Bathurst Manor / Wilson Heights / Downsview North   
3                                    Bayview Village   
4                 Bedford Park / Lawrence Manor East   

       1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0  Latin American Restaurant                Lounge        Breakfast Spot   
1                Pizza Place                   Gym          Skating Rink   
2                       Bank           Coffee Shop              Pharmacy   
3        Japanese Restaurant                  Café                  Bank   
4                Coffee Shop           Pizza Place            Restaurant   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0      Doner Restaurant    Dim Sum Restaurant                 Diner   
1              Pharmacy                   Pub        Sandwich Place   
2           Supermarket         Deli / Bodega      Sushi Restaurant   
3    Chinese Restaurant          Dessert Shop                 Diner   
4        Sandwich Place    Italian Restaurant      Sushi Restaurant   

     7th Most Common Venue      8th Most Common Venue 9th Most Common Venue  \
0           Discount Store        Distribution Center               Dog Run   
1              Coffee Shop            Airport Service           College Gym   
2               Restaurant  Middle Eastern Restaurant     Convenience Store   
3           Discount Store        Distribution Center               Dog Run   
4  Comfort Food Restaurant             Cosmetics Shop              Pharmacy   

  10th Most Common Venue  
0              Drugstore  
1   Ethiopian Restaurant  
2    Fried Chicken Joint  
3          Women's Store  
4                   Café

Good, now we have the dataframe to cluster
# Clustering
we will cluster them into 3 mutually exclusive clusters

In [90]:
# set number of clusters
kclusters = 3

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

Joining both datframes with location and venues

In [101]:
# add clustering labels
#neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = Toronto_Neighborhood


# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.merge(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
toronto_merged.head()

Postal code           Borough                                  Neighborhood  \
0         M3A        North York                                     Parkwoods   
1         M4A        North York                              Victoria Village   
2         M5A  Downtown Toronto                    Regent Park / Harbourfront   
3         M6A        North York             Lawrence Manor / Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656               1                  Park   
1  43.725882 -79.315572               0     French Restaurant   
2  43.654260 -79.360636               0           Coffee Shop   
3  43.718518 -79.464763               0        Clothing Store   
4  43.662301 -79.389494               0           Coffee Shop   

   2nd Most Common Venue       3rd Most Common Venue 4th Most Common Venue  \
0      Food & Drink Shop  Construction & Landscaping     Electronics Store   
1  Portuguese Restaurant                 Coffee Shop           Pizza Place   
2                    Pub                        Park                Bakery   
3          Women's Store      Furniture / Home Store           Coffee Shop   
4       Sushi Restaurant                       Diner        Discount Store   

         5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0  Eastern European Restaurant             Drugstore            Donut Shop   
1                 Intersection          Hockey Arena      Department Store   
2                      Theater            Restaurant                  Café   
3           Miscellaneous Shop           Event Space              Boutique   
4                         Park    Mexican Restaurant             Juice Bar   

  8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant          Deli / Bodega                Dog Run  
1          Dessert Shop     Dim Sum Restaurant                  Diner  
2        Breakfast Spot                    Spa         Ice Cream Shop  
3     Accessories Store  Vietnamese Restaurant       Department Store  
4   Japanese Restaurant     Italian Restaurant             Hobby Shop

In [106]:
# create map
map_clusters = folium.Map(location=[43.7538, -79.3832], zoom_start=10)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Neigborhoods in Cluster 0

In [118]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0,['Neighborhood','1st Most Common Venue','2nd Most Common Venue','3rd Most Common Venue','4th Most Common Venue','5th Most Common Venue','6th Most Common Venue','7th Most Common Venue','8th Most Common Venue','9th Most Common Venue','10th Most Common Venue']]

Neighborhood  \
1                                    Victoria Village   
2                          Regent Park / Harbourfront   
3                   Lawrence Manor / Lawrence Heights   
4        Queen's Park / Ontario Provincial Government   
5                                     Malvern / Rouge   
6                                           Don Mills   
7                                           Don Mills   
8                    Parkview Hill / Woodbine Gardens   
9                            Garden District, Ryerson   
10                                          Glencairn   
12                                   Woodbine Heights   
13                                     St. James Town   
14                                 Humewood-Cedarvale   
15  Eringate / Bloordale Gardens / Old Burnhamthor...   
16                Guildwood / Morningside / West Hill   
17                                        The Beaches   
18                                        Berczy Park   
20                                             Woburn   
21                                            Leaside   
22                                 Central Bay Street   
23                                           Christie   
24                                          Cedarbrae   
25                                  Hillcrest Village   
26  Bathurst Manor / Wilson Heights / Downsview North   
27                                   Thorncliffe Park   
28                         Richmond / Adelaide / King   
29                      Dufferin / Dovercourt Village   
31                     Fairview / Henry Farm / Oriole   
32                   Northwood Park / York University   
34  Harbourfront East / Union Station / Toronto Is...   
35                          Little Portugal / Trinity   
36      Kennedy Park / Ionview / East Birchmount Park   
37                                    Bayview Village   
38                                          Downsview   
39                                          Downsview   
40                                          Downsview   
41                                          Downsview   
42                      The Danforth West / Riverdale   
43          Toronto Dominion Centre / Design Exchange   
44     Brockton / Parkdale Village / Exhibition Place   
45                  Golden Mile / Clairlea / Oakridge   
46                    India Bazaar / The Beaches West   
47                    Commerce Court / Victoria Hotel   
49                                      Humber Summit   
50  Cliffside / Cliffcrest / Scarborough Village West   
51                                    Studio District   
52                 Bedford Park / Lawrence Manor East   
53  Del Ray / Mount Dennis / Keelsdale and Silvert...   
54                                  Humberlea / Emery   
55                       Birch Cliff / Cliffside West   
56                                         Willowdale   
57                                         Willowdale   
59                                           Roselawn   
60                     Runnymede / The Junction North   
62  Dorset Park / Wexford Heights / Scarborough To...   
64                                   Davisville North   
66                     High Park / The Junction South   
67                                          Westmount   
68                                 Wexford / Maryvale   
69                                 North Toronto West   
70              The Annex / North Midtown / Yorkville   
71                            Parkdale / Roncesvalles   
72              Canada Post Gateway Processing Centre   
74                                          Agincourt   
75                                         Davisville   
76                    University of Toronto / Harbord   
77                                Runnymede / Swansea   
78          Clarks Corners / Tam O'Shanter / Sullivan   
80        Kensington Market / Chinatown / Grange Park   
82  Summerhill West / Rathnelly / South Hill / For.

## Neigborhoods in Cluster 1

In [119]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1,['Neighborhood','1st Most Common Venue','2nd Most Common Venue','3rd Most Common Venue','4th Most Common Venue','5th Most Common Venue','6th Most Common Venue','7th Most Common Venue','8th Most Common Venue','9th Most Common Venue','10th Most Common Venue']]

Neighborhood  \
0                                           Parkwoods   
19                                Caledonia-Fairbanks   
30                                Scarborough Village   
48         North Park / Maple Leaf Park / Upwood Park   
58                                      Lawrence Park   
65                           Forest Hill North & West   
73  Kingsview Village / St. Phillips / Martin Grov...   
79                       Moore Park / Summerhill East   
81  Milliken / Agincourt North / Steeles East / L'...   
87                                           Rosedale   
93    The Kingsway / Montgomery Road / Old Mill North   

      1st Most Common Venue 2nd Most Common Venue       3rd Most Common Venue  \
0                      Park     Food & Drink Shop  Construction & Landscaping   
19                     Park         Women's Store                        Pool   
30  Health & Beauty Service            Playground               Women's Store   
48                     Park      Basketball Court                      Bakery   
58                     Park           Swim School                    Bus Line   
65                     Park                 Trail            Sushi Restaurant   
73                     Park           Pizza Place              Sandwich Place   
79                     Park          Tennis Court                  Playground   
81                     Park            Playground         Distribution Center   
87                     Park                 Trail                  Playground   
93                     Park                 River                  Smoke Shop   

         4th Most Common Venue        5th Most Common Venue  \
0            Electronics Store  Eastern European Restaurant   
19        Ethiopian Restaurant            Electronics Store   
30                     Dog Run             Department Store   
48  Construction & Landscaping           Dim Sum Restaurant   
58         Distribution Center                 Dessert Shop   
65                    Bus Line                Jewelry Store   
73           Mobile Phone Shop         Colombian Restaurant   
79                  Restaurant               Discount Store   
81               Deli / Bodega             Department Store   
87               Deli / Bodega             Department Store   
93          College Rec Center                  College Gym   

          6th Most Common Venue        7th Most Common Venue  \
0                     Drugstore                   Donut Shop   
19  Eastern European Restaurant                    Drugstore   
30                 Dessert Shop           Dim Sum Restaurant   
48                        Diner               Discount Store   
58           Dim Sum Restaurant                        Diner   
65           Dim Sum Restaurant                        Diner   
73              College Stadium  Eastern European Restaurant   
79                Deli / Bodega             Department Store   
81                 Dessert Shop           Dim Sum Restaurant   
87                 Dessert Shop           Dim Sum Restaurant   
93            Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0       Doner Restaurant         Deli / Bodega                Dog Run  
19            Donut Shop      Doner Restaurant                Dog Run  
30                 Diner        Discount Store    Distribution Center  
48   Distribution Center               Dog Run       Doner Restaurant  
58        Discount Store               Dog Run          Deli / Bodega  
65        Discount Store   Distribution Center                Dog Run  
73             Drugstore            Donut Shop       Doner Restaurant  
79          Dessert Shop    Dim Sum Restaurant                  Diner  
81                 Diner        Discount Store          Women's Store  
87                 Diner        Discount Store    Distribution Center  
93             Drugstore            Donut Shop       Doner Restau

## Neigborhoods in Cluster 2

In [120]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2,['Neighborhood','1st Most Common Venue','2nd Most Common Venue','3rd Most Common Venue','4th Most Common Venue','5th Most Common Venue','6th Most Common Venue','7th Most Common Venue','8th Most Common Venue','9th Most Common Venue','10th Most Common Venue']]

Neighborhood 1st Most Common Venue  \
11  Rouge Hill / Port Union / Highland Creek                   Bar   
33                              East Toronto                  Park   
61                                    Weston     Convenience Store   
63                           York Mills West                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
11         Women's Store      Doner Restaurant    Dim Sum Restaurant   
33     Convenience Store      Department Store          Dessert Shop   
61      Doner Restaurant          Dessert Shop    Dim Sum Restaurant   
63                  Bank                   Bar     Convenience Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
11                 Diner        Discount Store   Distribution Center   
33    Dim Sum Restaurant                 Diner        Discount Store   
61                 Diner        Discount Store   Distribution Center   
63               Dog Run    Dim Sum Restaurant                 Diner   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
11               Dog Run            Donut Shop         Farmers Market  
33   Distribution Center               Dog Run           Dance Studio  
61               Dog Run         Women's Store          Deli / Bodega  
63        Discount Store   Distribution Center       Doner Restaurant